libraries/Microsoft.Bot.Builder/StreamingRequestHandler.cs

```diff
@@ -0,0 +1,220 @@
+        /// </summary>
+        /// <param name="request">A ReceiveRequest from the connected channel.</param>
+        /// <param name="logger">Optional logger used to log request information and error details.</param>
+        /// <param name="context">Optional context to operate within. Unused in bot implementation.</param>
+        /// /// <param name="cancellationToken">Optional cancellation token.</param>
+        /// <returns>A response created by the BotAdapter to be sent to the client that originated the request.</returns>
+        public override async Task<StreamingResponse> ProcessRequestAsync(ReceiveRequest request, ILogger<RequestHandler> logger, object context = null, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            logger = logger ?? NullLogger<RequestHandler>.Instance;
+            var response = new StreamingResponse();
+
+            if (request == null || string.IsNullOrEmpty(request.Verb) || string.IsNullOrEmpty(request.Path))
+            {
+                response.StatusCode = (int)HttpStatusCode.BadRequest;
+                logger.LogInformation("Request missing verb and/or path.");
+
+                return response;
+            }
+
+            if (string.Equals(request.Verb, StreamingRequest.GET, StringComparison.InvariantCultureIgnoreCase) &&
+                         string.Equals(request.Path, "/api/version", StringComparison.InvariantCultureIgnoreCase))
+            {
+                response.StatusCode = (int)HttpStatusCode.OK;
+                response.SetBody(new VersionInfo() { UserAgent = _userAgent });
+
+                return response;
+            }
+
+            if (string.Equals(request.Verb, StreamingRequest.POST, StringComparison.InvariantCultureIgnoreCase) &&
+                         string.Equals(request.Path, "/api/messages", StringComparison.InvariantCultureIgnoreCase))
+            {
+                return await ProcessStreamingRequestAsync(request, response, logger, cancellationToken).ConfigureAwait(false);
+            }
+
+            response.StatusCode = (int)HttpStatusCode.NotFound;
+            logger.LogInformation($"Unknown verb and path: {request.Verb} {request.Path}");
+
+            return response;
+        }
+
+        /// <summary>
+        /// Build and return versioning information used for telemetry, including:
+        /// The Schema version is 3.1, put into the Microsoft-BotFramework header,
+        /// Protocol Extension Info,
+        /// The Client SDK Version
+        ///  https://github.com/Microsoft/botbuilder-dotnet/blob/d342cd66d159a023ac435aec0fdf791f93118f5f/doc/UserAgents.md,
+        /// Additional Info.
+        /// https://github.com/Microsoft/botbuilder-dotnet/blob/d342cd66d159a023ac435aec0fdf791f93118f5f/doc/UserAgents.md.
+        /// </summary>
+        /// <returns>A string containing versioning information.</returns>
+        private static string GetUserAgent() =>
+            string.Format(
```

Comment:
P1 - On every request, you're creating a new ConnectorClient. This seems like a mistake. 

Cache the value created for the Agent, and only every create it one time.

Reply:
The value for the Agent is created by the CTOR for the handler and cached for the lifespan of the handler, which is the lifespan of the connection. Each connection (which should be one per origin, but at worst one per conversation) use the same instance of the handler for all streaming requests. 

Within the Streaming Extensions library is a ProtocolAdapter that hands up incoming messages to the persisted request handler bound to the connection.